Issue #4029: Add a compare user layout visibility condition.

jenlampton opened the pull request:
Fixes https://github.com/backdrop/backdrop-issues/issues/4029

jenlampton commented:
Closing for new sandbox

jenlampton commented:
New sandbox?

klonos commented on core/modules/layout/includes/layout.layout.inc:
```diff
@@ -148,6 +148,23 @@ function layout_layout_access_info() {
+    // used in the $contexts array passed to the access class methods. The type
+    // references a context provided by hook_layout_context_info().
+    'required contexts' => array(
+      'user' => 'user',
```
Why just `user` instead of `user1` here? (same as bellow)

klonos commented on core/modules/layout/plugins/access/user_compare_layout_access.inc:
```diff
@@ -0,0 +1,70 @@
+    parent::form($form, $form_state);
+    $form['helptext'] = array(
+      '#type' => 'markup',
+      '#value' => '<div>' . t('Compare the two user accounts. For example, for a user to view their own profile, choose "logged in user" and "user being viewed" and say "grant access if equal".') . '</div>',
```
nit: we say "equal" here, but `$form['equality']['#options'][1]` below is using the word "same".

jenlampton replied:
addressed in the latest PR.

klonos commented on core/modules/layout/plugins/access/user_compare_layout_access.inc:
```diff
@@ -0,0 +1,70 @@
+   */
+  function summary($contexts = array()) {
+    if (empty($this->contexts) || count($this->contexts) != 2 || empty($this->contexts['user1']->data) || empty($this->contexts['user2']->data)) {
+      return t('Comparing two User accounts');
```
```suggestion
      return t('Comparing two user accounts');
```

klonos commented on core/modules/layout/plugins/access/user_compare_layout_access.inc:
```diff
@@ -0,0 +1,84 @@
+    );
+    $form['equality'] = array(
+      '#type' => 'radios',
+      '#title' => t('Grant access if the two are'),
```
We aren't granting access per se. How about since this is a visibility condition, we say "Show when..." instead?

Also, I think I preferred it better when the title was saying "if the two user accounts are", and then the options were simply "The same"/"Different"

jenlampton replied:
I like `Display when` or `Display if`

jenlampton replied:
Addressed in the latest PR.

klonos commented on core/modules/layout/includes/layout.layout.inc:
```diff
@@ -160,7 +160,24 @@ function layout_layout_access_info() {
+  );
+  $info['user_compare'] = array(
+    'title' => t('User: Compare'),
+    'description' => t('Compare two user accounts (logged-in user account vs user account being viewed, for example)'),
```
```suggestion
    'description' => t('Compare two user accounts (logged-in user account vs user account being viewed, for example)'),
```
```suggestion
    'description' => t('Compare two user accounts (logged-in user account vs user account being viewed, for example).'),
```

klonos commented on core/modules/layout/includes/layout.layout.inc:
```diff
@@ -160,7 +160,24 @@ function layout_layout_access_info() {
+    ),
+  );
+  $info['user_compare'] = array(
+    'title' => t('User: Compare'),
```
```suggestion
    'title' => t('User: Compare'),
```
```suggestion
    'title' => t('User account: Compare'),
```

klonos commented on core/modules/layout/plugins/access/user_compare_layout_access.inc:
```diff
@@ -0,0 +1,84 @@
+    parent::form($form, $form_state);
+    $form['helptext'] = array(
+      '#type' => 'markup',
+      '#value' => '<div>' . t('Compare two user accounts (logged-in account and account being viewed, for example)') . '</div>',
```
```suggestion
      '#value' => '<div>' . t('Compare two user accounts (logged-in account and account being viewed, for example).') . '</div>',
```

klonos replied:
...also, should we be using `'#type' => 'help'` and `'#markup' => ...` (without the `<div>`s) here?

Where is this help text meant to be shown?

jenlampton commented:
I'm unable to see the failures here, so I've rebased and pushed an update to see if I can get them to fail again :)

quicksketch commented on core/modules/layout/plugins/access/user_compare_layout_access.inc:
```diff
@@ -0,0 +1,93 @@
+<?php
+/**
+ * @file
+ * Plugin to provide access if two user account contexts are the same.
+ */
+class UserCompareLayoutAccess extends LayoutAccess {
```
I see the silly redundancy here with PHPCS requesting both a `@file` docblock and a class docblock. But I could see this being necessary in the event that namespaces/use statements are defined. In which case there would be lines between the `@file` docblock and the class docblock.

This was touched on in https://github.com/backdrop-ops/phpcs/issues/11

I'd suggest the `@file` docblock should simply be `Contains UserCompareLayoutAccess class.`

quicksketch commented on core/modules/layout/plugins/access/user_compare_layout_access.inc:
```diff
@@ -0,0 +1,97 @@
+      }
+    }
+
+    $comparison = !empty($this->settings['equality']) ? 'is the same as' : 'is different from';
+    return t('@id2 @comp the Logged-in user account.', array(
```
This needs to be separated out as two separate `t()` calls. With this approach `is the same as` and `is different from` is not translatable, and should be included as part of the `t()` string.

quicksketch commented on core/modules/layout/tests/layout.test:
```diff
@@ -1523,7 +1596,7 @@ class LayoutInterfaceTest extends BackdropWebTestCase {
     $this->assertEqual($result, $second_layout_name, 'Config created for second layout.');
     $result = config_get('layout.menu_item.' . $second_layout_name, 'name');
-    $this->assertNull($result, 'No config created for second layout menu item (reuses the existing one).');
+    $this->assertNULL($result, 'No config created for second layout menu item (reuses the existing one).');
```
These renamings should be reverted. The method is correctly named `assertNull`. I'm guessing this was accidental in a find/replace all.

quicksketch replied:
```suggestion
    $this->assertNull($result, 'No config created for second layout menu item (reuses the existing one).');
```

jenlampton commented:
@quicksketch  I've reverted the last "cleanup" commit, but left the [Merge branch '1.x' into 4029-layout-user-compare](https://github.com/backdrop/backdrop/pull/2869/commits/766e08ff486c5960f77d54d48696189e9075e857), and also rebased off the latest 1.x.

argiepiano commented on core/modules/layout/plugins/access/user_compare_layout_access.inc:
```diff
@@ -0,0 +1,97 @@
+        $labels[$key] = t($message, $substitutions);
+      }
+      else {
+        $labels[$key] = t('An unknown user account');
```
I find "unknown user account" a bit confusing. Can this be "A user account from the Layout context"

quicksketch replied:
I don't think this should ever happen normally? But you're right it's not unknown to the person that set the context, it's just unknown to the compare plugin what kind of user it is. Since it shouldn't occur, any label is probably fine so I don't think we need to waffle on it. I'll change to your suggested string.

quicksketch replied:
```suggestion
        $labels[$key] = t('A user account from the layout contexts');
```

quicksketch commented on core/modules/layout/tests/layout.test:
```diff
@@ -1547,7 +1620,7 @@ class LayoutInterfaceTest extends BackdropWebTestCase {
 
     $result = config_get('layout.menu_item.' . $second_layout_name, 'name');
-    $this->assertNull($result, 'Config for second menu item has been deleted.');
+    $this->assertNULL($result, 'Config for second menu item has been deleted.');
```
```suggestion
    $this->assertNull($result, 'Config for second menu item has been deleted.');
```

quicksketch commented on core/modules/layout/tests/layout.test:
```diff
@@ -2061,7 +2134,7 @@ class LayoutBlockTest extends BackdropWebTestCase {
     $this->assertTrue(strstr($classes_1, 'block-hero-no-image'));
     // Block 1: Confirm that no background image is present.
-    $this->assertNull($elements[0]['style'], 'No image present for 1st hero block.');
+    $this->assertNULL($elements[0]['style'], 'No image present for 1st hero block.');
```
```suggestion
    $this->assertNull($elements[0]['style'], 'No image present for 1st hero block.');
```

quicksketch commented on core/modules/layout/tests/layout.test:
```diff
@@ -3517,7 +3590,7 @@ class LayoutPathTest extends BackdropWebTestCase {
     $default_layout = $layouts['default'];
     $path = $default_layout->getPath(); // This will be NULL.
-    $this->assertNull($path, 'Default layout path is NULL.');
+    $this->assertNULL($path, 'Default layout path is NULL.');
```
```suggestion
    $this->assertNull($path, 'Default layout path is NULL.');
```

quicksketch commented on core/modules/layout/plugins/access/user_compare_layout_access.inc:
```diff
@@ -0,0 +1,97 @@
+    // Always assume the first context is the logged-in user account.
+    $form['contexts']['user1'] = array(
+      '#type' => 'value',
+      '#value' => 'current_user', 
```
```suggestion
      '#value' => 'current_user',
```

quicksketch commented:
Merged into 1.x for 1.27.0. Yay!
